File: KQXDrawer/KQXDrawer/LeftSlideViewController.m
//
//  LeftSlideViewController.m
//  KQXDrawer
//
//  Created by Kxx.xxQ 一生相伴 on 15/12/4.
//  Copyright © 2015年 asahi_kuang. All rights reserved.
//

#import "LeftSlideViewController.h"

@interface LeftSlideViewController ()
@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, assign) CGPoint tableViewCenter;
//头像
@property (nonatomic, strong) UIImageView *userIcon;


@end

@implementation LeftSlideViewController

#pragma mark - life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - methods
- (instancetype)initWithLeftViewController:(UIViewController *)leftVC withMainViewController:(UIViewController *)mainVC {
    self = [super init];
    if (self) {
        //
        self.speed = 0.3f;
        self.leftVc = leftVC;
        self.mainVc = mainVC;
        
        // 滑动手势
        self.panGes = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panGesHandler:)];
        [self.leftVc.view addGestureRecognizer:self.panGes];
        
        // 左边视图tableView获取
        for (UIView *view in self.leftVc.view.subviews) {
            if ([view isKindOfClass:[UITableView class]]) {
                self.tableView = (UITableView *) view;
            }
        }
        
        [self.tableView setFrame:REACT_MAKE(0, 165, REACT_WIDTH(FRAME) - SCREEN_WIDTH/2+65, REACT_HEIGHT(FRAME) - 150)];
        self.tableViewCenter = self.tableView.center;
        MAKE_SCALE(self.tableView, 0.1f, 0.1f);
        [self.tableView setCenter:CGPointMake(-REACT_WIDTH(self.tableView.frame), REACT_MINY(self.tableView.frame))];
        
        for (UIView *bgView in self.leftVc.view.subviews) {
            for (UIView *icon in bgView.subviews) {
                //
                if (icon.tag == ICON_TAG) {
                    self.userIcon = (UIImageView *) icon;
                }
            }
        }
        
        [self.userIcon setFrame:REACT_MAKE(50, 55, 100, 100)];
        MAKE_SCALE(self.userIcon, 0.1f, 0.1f);
        
        [self.view addSubview:self.leftVc.view];
        [self.view addSubview:self.mainVc.view];
    }
    return self;
}

- (void)openLeftViewController {
    [self.view addSubview:self.leftVc.view];
    [self.view addSubview:self.mainVc.view];
    
    [UIView animateWithDuration:0.8f delay:0 usingSpringWithDamping:0.7f initialSpringVelocity:0.1f options:UIViewAnimationOptionCurveLinear animations:^{
        //
        MAKE_SCALE(self.mainVc.view, 0.65f, 0.65f);
        [self.mainVc.view setCenter:CGPointMake(REACT_WIDTH(FRAME), REACT_HEIGHT(FRAME)/2)];
        
        MAKE_SCALE(self.tableView, 1, 1);
        [self.tableView setCenter:self.tableViewCenter];
        
        MAKE_SCALE(self.userIcon, 1, 1);
    } completion:^(BOOL finished) {
        //
    }];
    _isOpend = YES;
}

- (void)closeLeftViewController {
    [UIView animateWithDuration:0.8f delay:0 usingSpringWithDamping:0.7f initialSpringVelocity:0.1f options:UIViewAnimationOptionCurveLinear animations:^{
        //
        MAKE_TRANSLATION(self.mainVc.view, 0, 0);
        [self.mainVc.view setCenter:self.view.center];
        MAKE_SCALE(self.tableView, .1f, .1f);
        [self.tableView setCenter:CGPointMake(-REACT_WIDTH(self.tableView.frame), REACT_MINY(self.tableView.frame))];
        MAKE_SCALE(self.userIcon, 0.1f, 0.1f);
    } completion:^(BOOL finished) {
        //
        [self.leftVc.view removeFromSuperview];
    }];
    _isOpend = NO;
}

- (void)setPanGesEnabled:(BOOL)isEnable {
    [self.panGes setEnabled:isEnable];
}

#pragma mark - selector
- (void)panGesHandler:(UIPanGestureRecognizer *)pan {
    [self closeLeftViewController];
}


@end

File: KXXDrawer_Swift/KXXDrawer_Swift/LeftSlideViewController.swift
//
//  LeftSlideViewController.swift
//  KXXDrawer_Swift
//
//  Created by Kxx.xxQ 一生相伴 on 15/12/4.
//  Copyright © 2015年 asahi_kuang. All rights reserved.
//

import UIKit

class LeftSlideViewController: UIViewController {
    var leftVC   : UIViewController?
    var mainVC   : UIViewController?
    var isOpened : Bool?
    var tableView: UITableView?
    var tableViewCenter: CGPoint?
    var userIcon : UIImageView?
    
    // 构造方法，注意父类构造
    init(leftVC: UIViewController, mainVC: UIViewController) {
        self.leftVC = leftVC
        self.mainVC = mainVC
        
        super.init(nibName: nil, bundle: nil)
        
        
        
        for view in (self.leftVC?.view.subviews)! {
            if (view.isKindOfClass(UITableView.classForCoder()) == true) {
                self.tableView = (view as! UITableView)
            }
        }
        
        for imgView in (self.leftVC?.view.subviews)! {
            for icon in imgView.subviews {
                if icon.tag == 1991 {
                    self.userIcon = (icon as! UIImageView)
                }
            }
        }
        
        
        self.userIcon?.frame        = CGRectMake(75, 55, 100, 100)
        self.userIcon?.transform    = CGAffineTransformMakeScale(0.1, 0.1)

        self.tableView?.frame = CGRectMake(0, 180,CGRectGetWidth(self.view.frame) - CGRectGetWidth(self.view.frame)/2 + 65, CGRectGetHeight(self.view.frame) - 150)
        self.tableViewCenter = self.tableView?.center
        self.tableView?.transform = CGAffineTransformMakeScale(0.1, 0.1)
        self.tableView?.center = CGPointMake(-CGRectGetWidth((self.tableView?.frame)!), CGRectGetMinY((self.tableView?.frame)!))
        self.tableView?.transform   = CGAffineTransformMakeScale(0.1, 0.1)
        
        // 把两个控制器的视图添加上去
        self.view.addSubview(self.leftVC!.view)
        self.view.addSubview(self.mainVC!.view)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    // MARK: - life cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        self.createUI()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Create UI
    func createUI() {
        self.view.backgroundColor = UIColor.whiteColor()
        isOpened = false
    }
    // MARK: - Mothods
    func openLeftSlideViewController() {
        print("opened...");
        self.view.addSubview(self.leftVC!.view)
        self.view.addSubview(self.mainVC!.view)

        UIView.animateWithDuration(0.8, delay: 0, usingSpringWithDamping: 0.7, initialSpringVelocity: 0.6, options: UIViewAnimationOptions.CurveLinear, animations: { () -> Void in
            //
            self.mainVC?.view.transform = CGAffineTransformMakeScale(0.65, 0.65)
            self.mainVC?.view.center    = CGPointMake(CGRectGetWidth(self.view.frame), CGRectGetHeight(self.view.frame)/2)
            
            self.tableView?.transform   = CGAffineTransformMakeScale(1, 1)
            self.tableView?.center      = self.tableViewCenter!
            
            self.userIcon?.transform    = CGAffineTransformMakeScale(1, 1)
            
            }) { (Bool) -> Void in
                //
                self.isOpened = true
        }
    }
    
    func closeLeftSlideViewController() {
        print("closed...");
        UIView.animateWithDuration(0.8, delay: 0, usingSpringWithDamping: 0.7, initialSpringVelocity: 0.6, options: UIViewAnimationOptions.CurveLinear, animations: { () -> Void in
            //
            self.mainVC?.view.transform = CGAffineTransformMakeScale(1, 1)
            self.mainVC?.view.center    = CGPointMake(CGRectGetWidth(self.view.frame)/2, CGRectGetHeight(self.view.frame)/2)
            
            self.tableView?.transform = CGAffineTransformMakeScale(0.1, 0.1)
            self.tableView?.center = CGPointMake(-CGRectGetWidth((self.tableView?.frame)!), CGRectGetMinY((self.tableView?.frame)!))
            
            self.userIcon?.transform    = CGAffineTransformMakeScale(0.1, 0.1)

            }) { (Bool) -> Void in
                //
                self.leftVC?.view.removeFromSuperview()
                self.isOpened = false
        }
    }

}
